# Extracting by Label

### 1. Adapt

In [1]:
import pandas as pd
df1 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/General/CPC_with_labels.csv',skipinitialspace=True, chunksize=5000, header=0)
header = True
for chunk in df1:
    validid = chunk['Category'].str.contains('Adapt', regex=True)
    chunk = chunk[validid]
    chunk.to_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/Adapt1.csv', index=False, mode='a', header=header)
    header = False

In [2]:
df2 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/CPC_with_labels_2.csv',skipinitialspace=True, chunksize=5000, header=0)
header = True
for chunk in df2:
    validid = chunk['Category'].str.contains('Adapt', regex=True)
    chunk = chunk[validid]
    chunk.to_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/Adapt2.csv', index=False, mode='a', header=header)
    header = False

In [4]:
df3 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/Adapt1.csv', skipinitialspace=True, header=0)
df4 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/Adapt2.csv', skipinitialspace=True, header=0)
out = pd.concat([df3,df4])
out.to_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/Adapt.csv', index=False)

In [10]:
print ("concatenating " + str(df3['appln_id'].nunique()) + " unique ids from file 1 and " + str(df4['appln_id'].nunique()) + " and unique ids from file 2 to get " + str(out['appln_id'].nunique()) + " unique ids")
#~590,000 unique patents for Adapat category

concatenating 327402 unique ids from file 1 and 264595 and unique ids from file 2 to get 591997 unique ids


In [17]:
del df1, df2, df3, df4

In [ ]:
Others

In [21]:
Category = ['Build', 'Sequester', 'ICT', 'Geo', 'Hydro', 'Ocean', 'Solar-T', 'PV', 'Thermal-PV', 'Wind', 'Combust', 'Fusion', 'Fission', 'Electric', 'Biofuel', 'Waste', 'Store- B', 'Store - C', 'Store- T', 'Store- M', 'Hydrogen', 'Other-1', 'Other-2', 'Other-3', 'Processing', 'Transport', 'Wastewater', 'SmrtGrds', 'Other-4', 'Other-5', 'Nuclear', 'Non-Fossil', 'Enable','Other-6']
for c in Category:
    df1 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/General/CPC_with_labels.csv',skipinitialspace=True, chunksize=5000, header=0)
    header = True
    for chunk in df1:
        validid = chunk['Category'].str.contains(c, regex=True)
        chunk = chunk[validid]
        chunk.to_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/' + c +'1.csv', index=False, mode='a', header=header)
        header = False
    df2 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/General/CPC_with_labels_2.csv',skipinitialspace=True, chunksize=5000, header=0)
    header = True
    for chunk in df2:
        validid = chunk['Category'].str.contains(c, regex=True)
        chunk = chunk[validid]
        chunk.to_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/'+ c +'2.csv', index=False, mode='a', header=header)
        header = False
    df3 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/' +c+ '1.csv', skipinitialspace=True, header=0)
    df4 = pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/' +c+ '2.csv', skipinitialspace=True, header=0)
    out = pd.concat([df3,df4])
    out.to_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/' +c+ '.csv', index=False)
    print ("concatenating " + str(df3['appln_id'].nunique()) + " unique ids from file 1 and " + str(df4['appln_id'].nunique()) + " and unique ids from file 2 to get " + str(out['appln_id'].nunique()) + " unique ids")
    del df1, df2, df3, df4

concatenating 191276 unique ids from file 1 and 192832 and unique ids from file 2 to get 384108 unique ids
concatenating 24828 unique ids from file 1 and 20048 and unique ids from file 2 to get 44876 unique ids
concatenating 86299 unique ids from file 1 and 111194 and unique ids from file 2 to get 197493 unique ids
concatenating 5023 unique ids from file 1 and 4532 and unique ids from file 2 to get 9555 unique ids
concatenating 96947 unique ids from file 1 and 69017 and unique ids from file 2 to get 165964 unique ids
concatenating 12530 unique ids from file 1 and 11928 and unique ids from file 2 to get 24458 unique ids
concatenating 81656 unique ids from file 1 and 55862 and unique ids from file 2 to get 137518 unique ids
concatenating 103545 unique ids from file 1 and 140769 and unique ids from file 2 to get 244314 unique ids
concatenating 2168 unique ids from file 1 and 3020 and unique ids from file 2 to get 5188 unique ids
concatenating 61239 unique ids from file 1 and 75555 and uni

Adapt: 591,997 unique ids 

Build: 384,108 unique ids

Sequester: 44,876 unique ids

ICT: 197,493 unique ids

Geo: 9,555 unique ids

Hydro: 165,964 unique ids

Ocean: 24,458 unique ids

Solar-T: 137,518 unique ids

PV: 244,314 unique ids

Thermal-PV: 5,188 unique ids

Wind: 136,794 unique ids

Combust: 72,193 unique ids

Fusion: 7,123 unique ids

Fission: 35,425 unique ids

Electric: 54,905 unique ids

Biofuel: 69,976 unique ids

Waste: 407,554 unique ids

Store- B: 13 unique ids

Store - C: 37,054 unique ids

Store- T: 27,957 unique ids

Store- M: 5,763 unique ids

Hydrogen: 99,528 unique ids

Other-1: 6,059 unique ids

Other-2: 3 unique ids

Other-3: 9,817 unique ids

Processing: 1,000,415 unique ids

Transport: 784,595 unique ids

Wastewater: 362,185 unique ids

SmrtGrds: 75,885 unique ids

Other-4: 1,892,434 unique ids

Other-5: 3,190,434 unique ids

Nuclear: 68,959 unique ids

Non-Fossil: 41 unique ids

Enable: 97,886 unique ids

Other-6: 4,290 unique ids

~10.2 million unique ids (including overlap between categories)

In [28]:
#checking if all 8.5 million unique Y02 ids are retained
import pandas as pd
ids=pd.DataFrame()
Category = ['Adapt','Build', 'Sequester', 'ICT', 'Geo', 'Hydro', 'Ocean', 'Solar-T', 'PV', 'Thermal-PV', 'Wind', 'Combust', 'Fusion', 'Fission', 'Electric', 'Biofuel', 'Waste', 'Store- B', 'Store - C', 'Store- T', 'Store- M', 'Hydrogen', 'Other-1', 'Other-2', 'Other-3', 'Processing', 'Transport', 'Wastewater', 'SmrtGrds', 'Other-4', 'Other-5', 'Nuclear', 'Non-Fossil', 'Enable','Other-6']
for c in Category:
    df= pd.read_csv('/Volumes/NJ_4TB/PATSAT/MyFiles/Extras/' +c+ '.csv',skipinitialspace=True, chunksize=500000, header=0)
    header = True
    for chunk in df:
        ids=pd.concat([ids,(chunk["appln_id"])])

In [29]:
import numpy as np
ids.to_numpy().shape


(17241378, 1)

In [31]:
np.unique(ids.to_numpy()).shape

(8550374,)

8,550,374 (all ids are retained)